File: backend/src/domain/scheduling/domain.rs
#![allow(dead_code)]
use std::sync::Arc;

use crate::{
    datasource::{dbdao::dao::DBDao, scylladao::dao::ScyllaDao, vectordao::dao::VectorDao},
    domain::scheduling::task::Task,
    infra::registry::Registry,
};

#[derive(Clone)]
pub struct SchedulingDomain {
    pub db_dao: Arc<DBDao>,
    pub scylla_dao: Arc<ScyllaDao>,
    pub vector_dao: Arc<VectorDao>,
}

impl SchedulingDomain {
    pub fn new(r: &Registry) -> Self {
        Self {
            db_dao: r.db_dao.clone(),
            scylla_dao: r.scylla_dao.clone(),
            vector_dao: r.vector_dao.clone(),
        }
    }

    pub async fn get_task(&self, id: i64) -> Option<Task> {
        self.db_dao.get_task(id).await.ok().map(|record| Task {
            id: record.id,
            name: record.name,
            description: None,
            task_type: record.task_type,
            scheduler_id: record.scheduler_id,
            cron_expression: record.cron_expression,
            next_run_at: record.next_run_at,
            last_run_at: record.last_run_at,
            last_run_status: record.last_run_status,
            last_run_error: record.last_run_error,
            last_run_duration: record.last_run_duration,
            last_run_result: None,
            enabled: Some(record.enabled),
            created_at: record.created_at,
            updated_at: None,
        })
    }

    pub async fn search_tasks(
        &self,
        name: Option<String>,
        task_type: Option<String>,
        scheduler_id: Option<i64>,
        enabled: Option<bool>,
        page: i64,
        limit: i64,
    ) -> (Vec<Task>, i64) {
        match self
            .db_dao
            .search_tasks(name, task_type, scheduler_id, enabled, page, limit)
            .await
        {
            Ok(records) => {
                let tasks: Vec<Task> = records
                    .into_iter()
                    .map(|record| Task {
                        id: record.id,
                        name: record.name,
                        description: None,
                        task_type: record.task_type,
                        scheduler_id: record.scheduler_id,
                        cron_expression: record.cron_expression,
                        next_run_at: record.next_run_at,
                        last_run_at: record.last_run_at,
                        last_run_status: record.last_run_status,
                        last_run_error: record.last_run_error,
                        last_run_duration: record.last_run_duration,
                        last_run_result: None,
                        enabled: Some(record.enabled),
                        created_at: record.created_at,
                        updated_at: None,
                    })
                    .collect();
                let total = tasks.len() as i64;
                (tasks, total)
            }
            Err(_) => (Vec::new(), 0),
        }
    }

    pub async fn add_task(
        &self,
        name: String,
        task_type: String,
        scheduler_id: i64,
        cron_expression: Option<String>,
        enabled: bool,
    ) -> Result<i64, String> {
        self.db_dao
            .add_task(
                name.as_str(),
                task_type.as_str(),
                scheduler_id,
                cron_expression,
                enabled,
            )
            .await
            .map_err(|e| e.to_string())
    }

    pub async fn update_task(
        &self,
        id: i64,
        name: Option<String>,
        task_type: Option<String>,
        scheduler_id: Option<i64>,
        cron_expression: Option<String>,
        enabled: Option<bool>,
    ) -> Result<i64, String> {
        self.db_dao
            .update_task(id, name, task_type, scheduler_id, cron_expression, enabled)
            .await
            .map_err(|e| e.to_string())
    }

    pub async fn run_task(&self, id: i64) -> Result<i64, String> {
        self.db_dao.run_task(id).await.map_err(|e| e.to_string())
    }

    pub async fn stop_task(&self, id: i64) -> Result<i64, String> {
        self.db_dao.stop_task(id).await.map_err(|e| e.to_string())
    }

    pub async fn restart_task(&self, id: i64) -> Result<i64, String> {
        self.db_dao
            .restart_task(id)
            .await
            .map_err(|e| e.to_string())
    }
}

File: backend/domain/scheduling/domain.go
package scheduling

import (
	"backend/datasource/dbdao"
	"context"
	"errors"
)

type SchedulingDomain struct {
	DB *dbdao.DB
}

func NewSchedulingDomain(db *dbdao.DB) *SchedulingDomain {
	return &SchedulingDomain{DB: db}
}

func (d *SchedulingDomain) GetTask(ctx context.Context, id int64) (*Task, error) {
	t, err := d.DB.GetTask(ctx, id)
	if err != nil {
		return nil, err
	}
	return &Task{
		ID:             t.ID,
		Name:           t.Name,
		Description:    t.Description,
		TaskType:       t.TaskType,
		SchedulerID:    t.SchedulerID,
		CronExpression: t.CronExpression,
		Enabled:        t.Enabled,
		NextRunAt:      t.NextRunAt,
		LastRunAt:      t.LastRunAt,
		LastRunStatus:  t.LastRunStatus,
		LastRunError:   t.LastRunError,
		CreatedAt:      &t.CreatedAt,
		UpdatedAt:      &t.UpdatedAt,
	}, nil
}

func (d *SchedulingDomain) SearchTasks(ctx context.Context, name, taskType string, schedulerID int64, enabled *bool, page, limit int64) ([]*Task, int64, error) {
	if page < 1 {
		page = 1
	}
	if limit < 1 || limit > 100 {
		limit = 20
	}
	ts, total, err := d.DB.SearchTasks(ctx, name, taskType, schedulerID, enabled, page, limit)
	if err != nil {
		return nil, 0, err
	}
	result := make([]*Task, len(ts))
	for i, t := range ts {
		result[i] = &Task{
			ID: t.ID, Name: t.Name, Description: t.Description, TaskType: t.TaskType, SchedulerID: t.SchedulerID,
			CronExpression: t.CronExpression, Enabled: t.Enabled, NextRunAt: t.NextRunAt, LastRunAt: t.LastRunAt,
			LastRunStatus: t.LastRunStatus, LastRunError: t.LastRunError, CreatedAt: &t.CreatedAt, UpdatedAt: &t.UpdatedAt,
		}
	}
	return result, total, nil
}

func (d *SchedulingDomain) AddTask(ctx context.Context, name, taskType string, schedulerID int64, cronExpression string, enabled bool) (*Task, error) {
	if name == "" || taskType == "" {
		return nil, errors.New("missing required fields")
	}
	t := &dbdao.Task{Name: name, TaskType: taskType, SchedulerID: schedulerID, CronExpression: cronExpression, Enabled: enabled}
	created, err := d.DB.CreateTask(ctx, t)
	if err != nil {
		return nil, err
	}
	return &Task{ID: created.ID, Name: created.Name, TaskType: created.TaskType, SchedulerID: created.SchedulerID, CronExpression: created.CronExpression, Enabled: created.Enabled, CreatedAt: &created.CreatedAt, UpdatedAt: &created.UpdatedAt}, nil
}

func (d *SchedulingDomain) UpdateTask(ctx context.Context, id int64, name, taskType, cronExpression *string, schedulerID *int64, enabled *bool) (*Task, error) {
	if id <= 0 {
		return nil, errors.New("invalid id")
	}
	t := &dbdao.Task{ID: id}
	if name != nil {
		t.Name = *name
	}
	if taskType != nil {
		t.TaskType = *taskType
	}
	if schedulerID != nil {
		t.SchedulerID = *schedulerID
	}
	if cronExpression != nil {
		t.CronExpression = *cronExpression
	}
	if enabled != nil {
		t.Enabled = *enabled
	}
	updated, err := d.DB.UpdateTask(ctx, t)
	if err != nil {
		return nil, err
	}
	return &Task{ID: updated.ID, Name: updated.Name, TaskType: updated.TaskType, SchedulerID: updated.SchedulerID, CronExpression: updated.CronExpression, Enabled: updated.Enabled, CreatedAt: &updated.CreatedAt, UpdatedAt: &updated.UpdatedAt}, nil
}

func (d *SchedulingDomain) RunTask(ctx context.Context, id int64) (int64, error) {
	if id <= 0 {
		return 0, errors.New("invalid id")
	}
	return id, d.DB.RunTask(ctx, id)
}

func (d *SchedulingDomain) StopTask(ctx context.Context, id int64) (int64, error) {
	if id <= 0 {
		return 0, errors.New("invalid id")
	}
	return id, d.DB.StopTask(ctx, id)
}

func (d *SchedulingDomain) RestartTask(ctx context.Context, id int64) (int64, error) {
	if id <= 0 {
		return 0, errors.New("invalid id")
	}
	return id, d.DB.RestartTask(ctx, id)
}
